### ISSUE TITLE
[API Proposal]: Consider support for IParsable<Uri> on System.Uri

### ISSUE THREAD

**Issue by javiercn** [dotnet/runtime#92285]
Edited by @MihaZupan on Sep 22nd to reflect the latest API proposal shape.

### Background and motivation

Since it was introduced, IParsable<T> is implemented by most framework types to support converting from a string to the type. In Blazor we support deserializing IParsable<T> types when processing form data, and this is one of the types we need to special case support for because they don't implement IParsable<Uri>

### API Proposal

```csharp
namespace System;

public class Uri : IParsable<Uri> // New interface with an explicit implementation
{
    static bool IParsable<Uri>.TryParse(string? s, IFormatProvider? provider, out Uri result);
    static Uri IParsable<Uri>.Parse(string s, IFormatProvider? provider);
}
```

The `IFormatProvider` will be ignored.
The implementation will delegate to `new Uri(s)`/`Uri.TryCreate(s)`.
The options used will be `UriKind = UriKind.RelativeOrAbsolute, AllowImplicitFilePaths = false`.

### API Usage

Allows `Uri` to use the same generic paths in existing implementations.
Same idea as the example given for [`IParsable<String>`](https://github.com/dotnet/runtime/issues/78842):
```c#
public static IEnumerable<T> Separate<T>(string source, char separator) where T : IParsable<T>
{
    return source.Split(separator)
        .Select(x => T.Parse(x, CultureInfo.InvariantCulture));
}
```

### Alternative Designs

- `ISpanParsable` and `IUtf8SpanParsable` were mentioned as potential additions, but were currently [rejected](https://github.com/dotnet/runtime/issues/92285#issuecomment-1730294745) on the grounds that they would not provide any performance advantages over `IParsable`.
- We could choose to allow `AllowImplicitFilePaths`.

**Comment by Sergio0694** [comment]
Shouldn't we also suggest it implementing `ISpanParsable<Uri>` while at it?

**Comment by MihaZupan** [comment]
This seems reasonable to me, but the interface implementation should be explicit (no reason to have both `Uri.TryCreate` and `Uri.TryParse` exposed).

We should consider having the behavior be such that it uses `AllowImplicitFilePaths = false` instead (once #59099 is implemented).

> Shouldn't we also suggest it implementing `ISpanParsable<Uri>` while at it?

Note that `Uri` has a `string OriginalString` property on it, so we'd have to materialize the string either way.

**Comment by neon-sunset** [comment]
> Shouldn't we also suggest it implementing `ISpanParsable<Uri>` while at it?

...and `IUtf8SpanParsable<Uri>` too.

**Comment by MihaZupan** [comment]
If the input is a valid Uri, we will be forced to allocate the original input string due to the existing API surface (`OriginalString`).
As such, being able to parse from a span would only benefit cases where it's common for the input not to be valid.
But since we're saying that the methods should work in terms of `UriKind.RelativeOrAbsolute`, that means practically all inputs will be valid (with extremely rare exceptions like a scheme with over 1024 characters).

In other words, having the overload that takes a span would never help perf-wise because of the existing API shape, at which point there's no real reason to include it.

> `IUtf8SpanParsable<Uri>`

What would be the use cases for this? Also, it runs into the same limitations as `ISpanParsable`.

**Comment by MihaZupan** [comment]
Marking this one as ready for review with only explicit implementations of `IParsable<Uri>`.

**Comment by tannergooding** [comment]
> What would be the use cases for this? Also, it runs into the same limitations as ISpanParsable.

Can avoid an additional transcoding step. Allows it to support parsing from generic contexts. Etc

It's basically all the same stuff everywhere. There are potential perf and usability opportunities. If those don't exist for the type, then we don't need to expose the additional interfaces. But, exposing them can often simplify things.

**Comment by stephentoub** [comment]
> Can avoid an additional transcoding step. 

Not if it has to create a string, anyway, in support of Uri.OriginalString.

**Comment by tannergooding** [comment]
> Not if it has to create a string, anyway, in support of Uri.OriginalString.

That depends on how it's implemented.

If you have a UTF8 span and only expose a `string` parse API, then for something like `TryParse` you have to first transcode the UTF8 span to a UTF16 string (also incurring an allocation) and then validate it's correct.

If you expose `TryParse(ROSpan<byte>)` then you can perform the transcoding and validation simultaneously and only allocate the backing string on success, which is effectively a memcpy operation at that point.

**Comment by stephentoub** [comment]
> and only allocate the backing string on success

This is why Miha stated "being able to parse from a span would only benefit cases where it's common for the input not to be valid".  Neither of us believe that to be common.

**Comment by bartonjs** [comment]
Looks good as proposed.

When discussing if the TryParse and Parse should be implicit or explicit, we feel that having both TryCreate and TryParse is sub-optimal.

```C#
namespace System;

public class Uri : IParsable<Uri> // New interface with an explicit implementation
{
    static bool IParsable<Uri>.TryParse(string? s, IFormatProvider? provider, out Uri result) =>
        TryCreate(uriString, RelativeOrAbsolute, NoImplicitFilePaths, out result);
    static Uri IParsable<Uri>.Parse(string s, IFormatProvider? provider) =>
        new Uri(s, RelativeOrAbsolute, NoImplicitFilePaths);
}
```

**Comment by KonH** [comment]
@javiercn @bartonjs 
I made first draft of proposed changed, please review - https://github.com/dotnet/runtime/pull/93479.
I'm not sure how to properly handle nullability in TryParse method, so decided to just create new instance (it looks little bit better than ! usage). 
And as I see NoImplicitFilePaths is not able to be implemented right now.
